api/utils/rotatewriter/rotate_writer.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package rotatewriter
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+)
+
+type RotateWriter interface {
+	Start() error
+	Write(p []byte) (int, error)
+}
+
+type Writer struct {
+	dirPath         string
+	fileBaseName    string
+	maxFileSize     int64
+	maxNumFiles     int
+	currentFile     *os.File
+	currentSize     int64
+	numFilesRotated int
+}
+
+func (w *Writer) Start() error {
+	// Open the current log file
+	err := w.openNextFile()
+	if err != nil {
+		return fmt.Errorf("unable to open next file")
+	}
+
+	return nil
+}
+
+func (w *Writer) Write(p []byte) (int, error) {
+	if w.currentFile == nil {
+		return 0, io.ErrClosedPipe
+	}
+
+	// Rotate the log file if it exceeds the maximum file size
+	if w.currentSize+int64(len(p)) > w.maxFileSize {
+		err := w.rotateLogFile()
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// Write the log message to the current log file
+	n, err := w.currentFile.Write(p)
```

Comment:
Should we do?

```go
n, err := w.currentFile.Write(append(p, '\n'))
```

<img width="2087" alt="image" src="https://github.com/vechain/thor/assets/107671032/d4a201ab-f013-4977-be7e-8b748ccc9483">

Reply:
Damn, I thought I pushed that this morning.
We should defo append the \n, but if we do it in the api_request_logger, then the writer becomes agnostic of format.